book/src/rfcs/RFC0009-Updatable-Inputs.md

```diff
@@ -0,0 +1,63 @@
+# Updatable inputs
+
+## Metadata
+
+* Author: 1tgr
+* Date: 2021-06-17
+* Introduced in: https://github.com/salsa-rs/salsa/pull/1 (please update once you open your PR)
+
+## Summary
+
+- Update the value of an input query in place
+
+## Motivation
+
+- Adds an `update_some_input` function to input queries, which allows amendment of an input query's
+  value without cloning it.
+
+## User's guide
+
+Input queries in Salsa are defined in terms of a pair of functions:
+
+- `fn some_input(&self, key: u32) -> String`
+- `fn set_some_input(&self, key: u32, value: String)`
+
+Initial values are typically set when the database is instantiated, and are updated using a pattern
+like this:
+
+```rust
+let mut value = db.some_input(123);
+value.push_str("hello");
+db.set_some_input(123, value);
+```
+
+The `some_input` function works by fetching the value from the input storage and returning a clone.
+Cloning this value can be expensive.
+
+This proposal exposes a third function, which can be used to update the value on an input query in
+place, through a mutable reference to the underlying storage:
+
+- `fn update_some_input<F>(&self, key: u32, value_fn: F) where F: FnOnce(&mut String)`
```

Comment:
I think it should return `-> salsa::ValueChanged`, just like the other update function you proposed. It sometimes happens that the input value was the same as the new value you need.

Reply:
Ah, I see you addressed that point. We will have to make a new revision whatever happens, but I'm not sure why we should force the value to be assumed to have changed.

Reply:
Right, I think it's helpful for the closure to be able to change its mind by returning early and requesting that the value not be invalidated. Let's make these two new `update` functions consistent and update this one to return `salsa::ValueChanged`?

Reply:
> We will have to make a new revision whatever happens

So we would have the roll back the update in the `ValueChanged::False` case? In any case, this means decrementing the current revision number. But we may have also cancelled some queries via `SharedState::pending_revision`, which can't be un-cancelled.

Reply:
I think we would just always increment the revision -- but we can still avoid altering the "last changed revision" for this input, which means that revalidating derived queries should be cheap.